### DISCUSSION TITLE
Add extension support to extend nullability checks in jdt compiler

### DISCUSSION THREAD

**Discussion by gayanper** [eclipse-jdt/eclipse.jdt.core#641] [Ideas]
Today when new nullability checks are added like 
- Java inbuilt utility classes such as Objects
- Utility class from external libraries such as Guava, Apache Commons etc
- Assertion libraries like JUnit, Assert4J

We need to modify the compiler code to support them. This is fine for utilities which are part of Java SDK, but for other libraries if we keep changing the compiler it will not be scalable. 

My suggestions are:
- Introduce extension points to support nullability checks for 3rd party libraries. This could be done by library authors them selfs.
- Depend on linting tools such as sonarlint by switching of JDT compiler nullability checks.

What should be best option we should choose for JDT ? Are there other suggestion for this ?

**Comment by stephan-herrmann** [comment]
For simple cases, obviously "eclipse external annotations" are the solution.

For more sophisticated support that cannot be expressed with existing annotations, there are a few architectural challenges that in the past prevented any work in such direction:
* We speak of extending the compiler (ecj), but this compiler is a plain Java application that cannot make use of Equinox extension mechanisms. 
  * You might argue that such extensions could be supported for some IDE mode, only. But people will kill us, if the IDE suggests to write code that cannot be built using your build tool of choice, which would typically call the standalone version of ecj.
* Next it would be non-trivial to actually define such an extension point. 
  * The simplest approach, to just ask an extension at certain points during analysis might require that the entire compiler be made API: at least AST, Bindings, which isn't feasible.
  * Short of that we'd have to prepare all relevant information and ask simple yes/no questions to the extension. This implies: the compiler defines the patterns it can consider, and the extension only decides if the pattern matches based on the actual code involved.

So, yes, the idea is nice, but the solution is all but trivial.
